File: src/MobileLmStudio/Services/LmStudioClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Options;
using MobileLmStudio.Models;

namespace MobileLmStudio.Services;

internal sealed class LmStudioClient
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    private readonly HttpClient _httpClient;
    private readonly IOptionsMonitor<AppOptions> _options;
    private readonly LmStudioConnectionService _connectionService;

    public LmStudioClient(HttpClient httpClient, IOptionsMonitor<AppOptions> options, LmStudioConnectionService connectionService)
    {
        _httpClient = httpClient;
        _options = options;
        _connectionService = connectionService;
        _httpClient.Timeout = Timeout.InfiniteTimeSpan;
    }

    public async Task<ModelCatalogResponse> GetModelsAsync(CancellationToken cancellationToken)
    {
        using var request = CreateRequest(HttpMethod.Get, "/api/v1/models");
        using var response = await _httpClient.SendAsync(request, cancellationToken);
        await EnsureSuccessAsync(response, cancellationToken);

        var payload = await response.Content.ReadFromJsonAsync<LmStudioModelCatalogResponse>(JsonOptions, cancellationToken)
            ?? new LmStudioModelCatalogResponse([]);

        var models = payload.Models
            .Select(model => new ModelDto(
                model.Key,
                model.DisplayName,
                model.Type,
                model.Publisher,
                model.Architecture,
                model.Quantization is null ? null : new QuantizationDto(model.Quantization.Name, model.Quantization.BitsPerWeight),
                model.SizeBytes,
                model.ParamsString,
                model.MaxContextLength,
                model.Format,
                model.LoadedInstances?.Select(instance => new LoadedModelInstanceDto(
                    instance.Id,
                    new ModelLoadConfigurationDto(
                        instance.Config.ContextLength,
                        instance.Config.EvalBatchSize,
                        instance.Config.Parallel,
                        instance.Config.FlashAttention,
                        instance.Config.NumExperts,
                        instance.Config.OffloadKvCacheToGpu))).ToArray() ?? [],
                model.Capabilities is null ? null : new ModelCapabilitiesDto(
                    model.Capabilities.Vision,
                    model.Capabilities.TrainedForToolUse,
                    model.Capabilities.Reasoning is null
                        ? null
                        : new ModelReasoningOptionsDto(model.Capabilities.Reasoning.AllowedOptions, model.Capabilities.Reasoning.Default)),
                model.Variants ?? [],
                model.SelectedVariant,
                model.Description))
            .ToArray();

        return new ModelCatalogResponse(models);
    }

    public async Task<LoadModelResponseDto> LoadModelAsync(ModelLoadRequest requestModel, CancellationToken cancellationToken)
    {
        using var request = CreateRequest(HttpMethod.Post, "/api/v1/models/load", new LmStudioLoadRequest
        {
            Model = requestModel.Model,
            ContextLength = requestModel.ContextLength,
            FlashAttention = requestModel.FlashAttention,
            EchoLoadConfig = true,
        });

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        await EnsureSuccessAsync(response, cancellationToken);

        var payload = await response.Content.ReadFromJsonAsync<LmStudioLoadResponse>(JsonOptions, cancellationToken)
            ?? throw new InvalidOperationException("LM Studio returned an empty load response.");

        return new LoadModelResponseDto(
            payload.Type,
            payload.InstanceId,
            payload.LoadTimeSeconds,
            payload.Status,
            payload.LoadConfig is null
                ? null
                : new ModelLoadConfigurationDto(
                    payload.LoadConfig.ContextLength,
                    payload.LoadConfig.EvalBatchSize,
                    payload.LoadConfig.Parallel,
                    payload.LoadConfig.FlashAttention,
                    payload.LoadConfig.NumExperts,
                    payload.LoadConfig.OffloadKvCacheToGpu));
    }

    public async Task<UnloadModelResponseDto> UnloadModelAsync(ModelUnloadRequest requestModel, CancellationToken cancellationToken)
    {
        using var request = CreateRequest(HttpMethod.Post, "/api/v1/models/unload", new LmStudioUnloadRequest
        {
            InstanceId = requestModel.InstanceId,
        });

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        await EnsureSuccessAsync(response, cancellationToken);

        var payload = await response.Content.ReadFromJsonAsync<LmStudioUnloadResponse>(JsonOptions, cancellationToken)
            ?? throw new InvalidOperationException("LM Studio returned an empty unload response.");

        return new UnloadModelResponseDto(payload.InstanceId);
    }

    public Task<HttpResponseMessage> StartChatStreamAsync(LmStudioChatRequest chatRequest, CancellationToken cancellationToken)
    {
        var request = CreateRequest(HttpMethod.Post, "/api/v1/chat", chatRequest, acceptEventStream: true);
        return _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
    }

    public async Task<string?> GenerateTitleAsync(string model, string input, CancellationToken cancellationToken)
    {
        var prompt = input.Length > 500 ? input[..500] : input;
        using var request = CreateRequest(HttpMethod.Post, "/api/v1/chat", new LmStudioChatRequest
        {
            Model = model,
            Input = prompt,
            SystemPrompt = "Generate a short chat title (5 words max). Respond with only the title, no quotes or extra text.",
            Stream = false,
            Store = false,
        });

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            return null;
        }

        var result = await response.Content.ReadFromJsonAsync<LmStudioChatResponse>(JsonOptions, cancellationToken);
        var title = result?.Output?.FirstOrDefault(o => o.Type == "message")?.Content?.Trim();
        if (string.IsNullOrWhiteSpace(title))
        {
            return null;
        }

        title = title.Trim('"', '\'');
        return title.Length > 60 ? $"{title[..57]}..." : title;
    }

    public async Task<string> ReadErrorAsync(HttpResponseMessage response, CancellationToken cancellationToken)
    {
        var content = await response.Content.ReadAsStringAsync(cancellationToken);
        return string.IsNullOrWhiteSpace(content)
            ? $"LM Studio returned {(int)response.StatusCode} {response.ReasonPhrase}."
            : content;
    }

    private HttpRequestMessage CreateRequest(HttpMethod method, string path, object? content = null, bool acceptEventStream = false)
    {
        var baseUrl = _connectionService.BaseUrl;
        var apiToken = _connectionService.ApiToken;
        var baseUri = new Uri(EnsureTrailingSlash(baseUrl), UriKind.Absolute);
        var request = new HttpRequestMessage(method, new Uri(baseUri, path));

        request.Headers.Accept.Clear();
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(acceptEventStream ? "text/event-stream" : "application/json"));

        if (!string.IsNullOrWhiteSpace(apiToken))
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiToken.Trim());
        }

        if (content is not null)
        {
            request.Content = new StringContent(JsonSerializer.Serialize(content, JsonOptions), Encoding.UTF8, "application/json");
        }

        return request;
    }

    private static string EnsureTrailingSlash(string baseUrl)
    {
        return baseUrl.EndsWith('/') ? baseUrl : $"{baseUrl}/";
    }

    private async Task EnsureSuccessAsync(HttpResponseMessage response, CancellationToken cancellationToken)
    {
        if (response.IsSuccessStatusCode)
        {
            return;
        }

        var details = await ReadErrorAsync(response, cancellationToken);
        throw new HttpRequestException(details, null, response.StatusCode);
    }
}

File: src/node/lm-studio-client.js
class LmStudioClient {
  constructor(getConnectionSettings) {
    this.getConnectionSettings = getConnectionSettings;
  }

  async getModels() {
    const response = await this.send("GET", "/api/v1/models");
    const payload = await response.json().catch(() => ({ models: [] }));
    const models = Array.isArray(payload.models) ? payload.models.map(mapModel) : [];
    return { models };
  }

  async loadModel(requestModel) {
    const response = await this.send("POST", "/api/v1/models/load", {
      model: requestModel.model,
      context_length: requestModel.contextLength,
      flash_attention: requestModel.flashAttention,
      echo_load_config: true
    });

    const payload = await response.json();
    return {
      type: payload.type,
      instanceId: payload.instance_id,
      loadTimeSeconds: payload.load_time_seconds,
      status: payload.status,
      loadConfig: payload.load_config ? mapLoadConfig(payload.load_config) : null
    };
  }

  async unloadModel(requestModel) {
    const response = await this.send("POST", "/api/v1/models/unload", {
      instance_id: requestModel.instanceId
    });

    const payload = await response.json();
    return {
      instanceId: payload.instance_id
    };
  }

  async startChatStream(chatRequest, signal) {
    return this.send("POST", "/api/v1/chat", chatRequest, {
      acceptEventStream: true,
      signal,
      allowFailure: true
    });
  }

  async generateTitle(model, input) {
    const prompt = String(input || "").slice(0, 500);
    const response = await this.send("POST", "/api/v1/chat", {
      model,
      input: prompt,
      system_prompt: "Generate a short chat title (5 words max). Respond with only the title, no quotes or extra text.",
      stream: false,
      store: false
    }, {
      allowFailure: true
    });

    if (!response.ok) {
      return null;
    }

    const payload = await response.json().catch(() => null);
    const output = Array.isArray(payload?.output) ? payload.output : [];
    const item = output.find(candidate => candidate?.type === "message" && candidate?.content);
    if (!item?.content) {
      return null;
    }

    const title = String(item.content).trim().replace(/^['"]|['"]$/g, "");
    return title.length > 60 ? `${title.slice(0, 57)}...` : title;
  }

  async generateAdaptiveMemory(model, options) {
    const response = await this.send("POST", "/api/v1/chat", {
      model,
      input: buildAdaptiveMemoryPrompt(options),
      system_prompt: [
        "Maintain a compact, durable memory profile for a chat user.",
        "Update the profile using the existing memory and the newly observed user messages.",
        "Keep only stable preferences, recurring interests, communication style, output preferences, and durable project context.",
        "Do not store secrets, one-off transient tasks, or speculative personal details.",
        `Return bullet points only and stay within ${Math.max(50, Number.parseInt(options?.maxWords, 10) || 500)} words.`
      ].join(" "),
      stream: false,
      store: false,
      temperature: 0,
      max_output_tokens: Math.min(Math.max((Number.parseInt(options?.maxWords, 10) || 500) * 5, 256), 2000)
    }, {
      allowFailure: true
    });

    if (!response.ok) {
      throw new Error(await this.readError(response));
    }

    const payload = await response.json().catch(() => null);
    const output = Array.isArray(payload?.output) ? payload.output : [];
    const item = output.find(candidate => candidate?.type === "message" && candidate?.content);
    if (!item?.content) {
      return String(options?.existingSummary || "").trim();
    }

    return truncateWords(String(item.content).trim(), Math.max(50, Number.parseInt(options?.maxWords, 10) || 500));
  }

  async readError(response) {
    const content = await response.text();
    if (!content.trim()) {
      return `LM Studio returned ${response.status} ${response.statusText}.`;
    }

    try {
      const payload = JSON.parse(content);
      const value = payload.detail || payload.error || payload.message || payload.title || content;
      return typeof value === "string" ? value : JSON.stringify(value);
    } catch {
      return content;
    }
  }

  async send(method, routePath, body, options = {}) {
    const url = new URL(routePath, ensureTrailingSlash(this.getConnectionSettings().baseUrl)).toString();
    const headers = {
      Accept: options.acceptEventStream ? "text/event-stream" : "application/json"
    };

    const apiToken = this.getConnectionSettings().apiToken;
    if (apiToken) {
      headers.Authorization = `Bearer ${apiToken.trim()}`;
    }

    if (body !== undefined) {
      headers["Content-Type"] = "application/json";
    }

    const response = await fetch(url, {
      method,
      headers,
      body: body === undefined ? undefined : JSON.stringify(body),
      signal: options.signal
    });

    if (!options.allowFailure && !response.ok) {
      throw new Error(await this.readError(response));
    }

    return response;
  }
}

function mapModel(model) {
  return {
    key: model.key,
    displayName: model.display_name,
    type: model.type,
    publisher: model.publisher,
    architecture: model.architecture ?? null,
    quantization: model.quantization
      ? {
          name: model.quantization.name ?? null,
          bitsPerWeight: model.quantization.bits_per_weight ?? null
        }
      : null,
    sizeBytes: model.size_bytes,
    paramsString: model.params_string ?? null,
    maxContextLength: model.max_context_length,
    format: model.format ?? null,
    loadedInstances: Array.isArray(model.loaded_instances)
      ? model.loaded_instances.map(instance => ({
          id: instance.id,
          config: mapLoadConfig(instance.config)
        }))
      : [],
    capabilities: model.capabilities
      ? {
          vision: Boolean(model.capabilities.vision),
          trainedForToolUse: Boolean(model.capabilities.trained_for_tool_use),
          reasoning: model.capabilities.reasoning
            ? {
                allowedOptions: Array.isArray(model.capabilities.reasoning.allowed_options) ? model.capabilities.reasoning.allowed_options : [],
                default: model.capabilities.reasoning.default
              }
            : null
        }
      : null,
    variants: Array.isArray(model.variants) ? model.variants : [],
    selectedVariant: model.selected_variant ?? null,
    description: model.description ?? null
  };
}

function mapLoadConfig(config) {
  return {
    contextLength: config.context_length,
    evalBatchSize: config.eval_batch_size ?? null,
    parallel: config.parallel ?? null,
    flashAttention: config.flash_attention ?? null,
    numExperts: config.num_experts ?? null,
    offloadKvCacheToGpu: config.offload_kv_cache_to_gpu ?? null
  };
}

function ensureTrailingSlash(baseUrl) {
  return String(baseUrl || "").endsWith("/") ? String(baseUrl) : `${baseUrl}/`;
}

function buildAdaptiveMemoryPrompt(options) {
  const existingSummary = String(options?.existingSummary || "").trim();
  const messages = Array.isArray(options?.messages) ? options.messages : [];
  const parts = [];

  parts.push("Existing memory profile:");
  parts.push(existingSummary || "- None yet.");
  parts.push("");
  parts.push("New user messages since the last review:");

  for (const message of messages) {
    const title = String(message?.chatTitle || "Untitled chat").trim();
    const createdAt = String(message?.createdAt || "").trim();
    const content = String(message?.content || "").trim();
    if (!content) {
      continue;
    }
    parts.push(`- [${createdAt || "unknown time"}] ${title}: ${content}`);
  }

  parts.push("");
  parts.push("Revise the memory profile so it stays compact, factual, and directly useful for future assistant behavior.");
  return parts.join("\n").trim();
}

function truncateWords(text, maxWords) {
  const words = String(text || "").trim().split(/\s+/).filter(Boolean);
  if (words.length <= maxWords) {
    return words.join(" ");
  }

  return words.slice(0, maxWords).join(" ");
}

module.exports = {
  LmStudioClient
};